### ISSUE TITLE
What to do if resolve takes couple of seconds?

### ISSUE THREAD

**Issue by SamVerschueren** [angular-ui/ui-router#1814]
Hi everyone

I actually stumbled across the resolve and onEnter features of ui-router recently and I thought I could make use of this to create middlewares for my controllers. For instance, if you have pages that can only be accessed if the user is signed in, the same middleware can be used to first check if the user has access to the page requested.

This is how I do it

```JavaScript
var middleware = angular.module('app.middlewares', []);

middleware.constant('middleware', {
    isAuthenticated: function() {
        return {
            resolve: {
                user: ['$q', 'authModel', '$timeout', function($q, authModel, $timeout) {
                    var deferred = $q.defer();

                    authModel.findUser(function(err, user) {
                        $timeout(function() {
                            deferred.resolve(user);
                        }, 5000);
                    });

                    return deferred.promise;
                }]
            },
            onEnter: ['$state', '$timeout', 'user', function($state, $timeout, user) {
                if(!user) {
                    $timeout(function() {
                        $state.go('auth');
                    });
                }
            }]
        };
    }
});
```

And this is my router
```JavaScript
var routes = angular.module('app.routes', ['ui.router', 'app.middlewares']);

routes.config(['$stateProvider', '$urlRouterProvider', 'middleware', function($stateProvider, $urlRouterProvider, middleware) {
    $urlRouterProvider.otherwise('/');

    $stateProvider
        .state('default', {
            abstract: true,
            templateUrl: 'app/views/includes/default.html',
            controller: 'Controller'
        })
        .state('titlebar', {
            abstract: true,
            parent: 'default',
            templateUrl: 'app/views/includes/titlebar.html',
            controller: 'TitleBarController'
        })
        .state('auth', {
            url: '/',
            parent: 'default',
            templateUrl: 'app/views/auth/signin.html',
            controller: 'AuthController'
        })
        .state('home', _.extend(middleware.isAuthenticated(), {
            url: '/home',
            templateUrl: 'app/views/home/index.html',
            parent: 'titlebar',
            controller: 'HomeController'
        }))
        .state('titlebar.settings', _.extend(middleware.isAuthenticated(), {
            url: '/settings',
            templateUrl: 'app/views/settings/index.html',
            controller: 'SettingsController'
        }));
}]);
```

As you can see, I use lodash to extend from the isAuthenticated middleware everytime. This way I can easily reuse the same pieces of code.

As you can see, I added a timeout in the ```isAuthenticated``` middleware to resolve the user only after 5 seconds. This is just for testing purposes offcourse to see how the application reacts when the process runs slower.

How can I actually handle this if the resolve process is slow (for instance because of internet connection, ...)? What happens now is that the application hangs for 5 seconds and then suddenly pops up the page requested. Is someone experiencing the same issues?

Is it good practice by using this 'middleware'-like approach or should/can it be done somewhere else? I want to avoid spoiling my controllers with all those checks over and over again.

Thanks in advance

**Comment by fabiosussetto** [comment]
For similar use cases, I've used this: http://chieffancypants.github.io/angular-loading-bar/
You can display a progress bar by injecting the bar dependency in your resolve function, and then show/hide the bar as your operation is done. With that, you can achieve something pretty similar to what Youtube does when you navigate through videos.

**Comment by SamVerschueren** [comment]
Thanks for the tip! Nice idea to do it this way.

**Comment by oldtimeguitarguy** [comment]
This might help? https://github.com/oldtimeguitarguy/angular-middleware
